pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,344 @@
+ * 8-Segments that go from CONSUMING to DROPPED or OFFLINE states stop being tracked so their delays do not cloud
+ *   delays of active partitions.
+ * 9-When a segment goes from CONSUMING to ONLINE or OFFLINE, we start ticking time for the corresponding partition.
+ *   If no consumption is noticed after a timeout, then we read ideal state to confirm the server still hosts the
+ *   partition. If not, we stop tracking the respective partition.
+ * 10-A timer thread is started by this object to track timeouts of partitions and rive the reading of their ideal
+ *    state.
+ *
+ *  The following diagram illustrates the object interactions with main external APIs
+ *
+ *     (CONSUMING -> ONLINE state change)
+ *             |
+ *      markPartition(partitionId)      |<-recordPinotConsumptionDelay()-{LLRealtimeSegmentDataManager(Partition 0}}
+ *            |                         |
+ * ___________V_________________________V_
+ * |           (Table X)                |<-recordPinotConsumptionDelay()-{LLRealtimeSegmentDataManager(Partition 1}}
+ * | MaximumPinotConsumptionDelayTracker|           ...
+ * |____________________________________|<-recordPinotConsumptionDelay()-{LLRealtimeSegmentDataManager (Partition n}}
+ *              ^                      ^
+ *              |                       \
+ *   tickInactivePartitions()    stopTrackingPinotConsumptionDelay(partitionId)
+ *    _________|__________                \
+ *   | TimerTrackingTask |          (CONSUMING -> DROPPED state change)
+ *   |___________________|          (CONSUMING -> OFFLINE state change)
+ *
+ */
+
+public class MaximumPinotConsumptionDelayTracker {
+  /*
+   * Class to handle timer thread that will track inactive partitions
+   */
+  private class TrackingTimerTask extends TimerTask {
+    private String _name;
+    private MaximumPinotConsumptionDelayTracker _tracker;
+
+    public TrackingTimerTask(String name, MaximumPinotConsumptionDelayTracker tracker) {
+      _name = name;
+      _tracker = tracker;
+    }
+
+    @Override
+    public void run() {
+      // tick inactive partitions every interval of time to keep tracked partitions up to date.
+      _tracker.tickInactivePartitions();
+    }
+  }
+
+  /*
+   * Class to keep a Pinot Consumption Delay measure and the time when the sample was taken (i.e. sample time)
+   * We will use the sample time to increase consumption delay when a partition stops consuming: the time
+   * difference between the sample time and current time will be added to the metric when read.
+   */
+  private class DelayMeasure {
+    public DelayMeasure(long t, long d) {
+      _delayMilliseconds = d;
+      _sampleTime = t;
+    }
+    public long _delayMilliseconds;
+    public long _sampleTime;
+  }
+
+  // HashMap used to store delay measures for all partitions active for the current table.
+  private HashMap<Long, DelayMeasure> _partitionToDelaySampleMap = new HashMap<>();
```

Comment:
The update method - `recordPinotConsumptionDelay` - is in the critical path of consumption and defining it as `synchronized` will be costly. I suggest use a concurrent hash map here so there's no need for synchronization.

Reply:
+1

Reply:
We now use concurrent hash maps and synchronized were removed

Reply:
refactored